### PR TITLE
[patch] change etag expire logic

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <i.can.feel.gravity@gmail.com>

### CODE REVIEW

File: policy/daemon.go [hunk 1]
```diff
@@ -321,7 +319,12 @@ func (p *policyd) fetchPolicy(ctx context.Context, domain string) (*SignedPolicy
 	eTag := res.Header.Get("ETag")
 	if eTag != "" {
 		glg.Debugf("Setting ETag %v for domain %s", eTag, domain)
-		p.etagCache.SetWithExpire(domain, &etagCache{eTag, sp}, p.etagExpTime)
+
+		dur := sp.SignedPolicyData.Expires.Sub(time.Now()) - p.expireMargin
```
Comment (kevindiu) [inline policy/daemon.go]: Please review this logic...

**Comment by WindzCUHK** [comment]
This closes #29 & closes #30

**Comment by kevindiu** [comment]
@WindzCUHK LGTM?

**Comment by WindzCUHK** [comment]
will add test case for https://github.com/yahoojapan/athenz-authorizer/pull/33/commits/fc498559c25549345f9053ea91850761d514a200 later

**Comment by WindzCUHK** [comment]
other issue...
1. WithPolicyExpiredDuration
	1. missing option in proxy
	1. inconsistent name

File: policy/daemon.go [hunk 2]
```diff
@@ -317,11 +316,28 @@ func (p *policyd) fetchPolicy(ctx context.Context, domain string) (*SignedPolicy
...
+		glg.Warnf("Warning: policy expiry not set, domain: %v", domain)
+		etagValidDur = p.etagExpTime - p.expireMargin
+	} else {
+		// if not set, sp.SignedPolicyData.Expires is nil
+		// if set, but invalid, sp.SignedPolicyData.Expires is Time{}
+		validDur := sp.SignedPolicyData.Expires.Time.Sub(fastime.Now())
+		if validDur <= p.expireMargin {
+			p.etagCache.Delete(domain)
+			glg.Errorf("Error verifying policy expiry, policy expiry: %v, valid duration: %v, valid margin: %v", sp.SignedPolicyData.Expires.Time, validDur, p.expireMargin)
+			return nil, false, errors.New("policy already expired")
```
Comment (kevindiu) [inline policy/daemon.go]: this logic is wrong, the policy is not expired yet but this error returned.

File: policy/daemon.go [hunk 3]
```diff
@@ -317,11 +316,28 @@ func (p *policyd) fetchPolicy(ctx context.Context, domain string) (*SignedPolicy
...
+	var etagValidDur time.Duration
+	if sp.SignedPolicyData.Expires == nil {
+		glg.Warnf("Warning: policy expiry not set, domain: %v", domain)
+		etagValidDur = p.etagExpTime - p.expireMargin
+	} else {
+		// if not set, sp.SignedPolicyData.Expires is nil
+		// if set, but invalid, sp.SignedPolicyData.Expires is Time{}
+		validDur := sp.SignedPolicyData.Expires.Time.Sub(fastime.Now())
+		if validDur <= p.expireMargin {
+			p.etagCache.Delete(domain)
```
Comment (kevindiu) [inline policy/daemon.go]: I think we do not need to delete the etag cache, just let it expire.

File: policy/daemon.go [hunk 4]
```diff
@@ -317,11 +316,28 @@ func (p *policyd) fetchPolicy(ctx context.Context, domain string) (*SignedPolicy
...
 
-	// set eTag cache
-	eTag := res.Header.Get("ETag")
-	if eTag != "" {
-		glg.Debugf("Setting ETag %v for domain %s", eTag, domain)
-		p.etagCache.SetWithExpire(domain, &etagCache{eTag, sp}, p.etagExpTime)
+	// verify policy expiry
+	var etagValidDur time.Duration
+	if sp.SignedPolicyData.Expires == nil {
+		glg.Warnf("Warning: policy expiry not set, domain: %v", domain)
```
Comment (kevindiu) [inline policy/daemon.go]: if expires is nil, maybe it is better to return error directly instead?

File: policy/daemon_fetchPolicy_test.go [hunk 5]
```diff
@@ -0,0 +1,598 @@
+/*
```
Comment (kevindiu) [inline policy/daemon_fetchPolicy_test.go]: why separate file?

File: policy/daemon_test.go [hunk 6]
```diff
@@ -773,6 +778,142 @@ func Test_policyd_CheckPolicy(t *testing.T) {
...
+				} else if err.Error() != tt.want.Error() {
+					t.Errorf("CheckPolicy error: err: %v, want: %v", err, tt.want)
+				}
+			}
+
+			// check runtime stack for go routine leak
+			time.Sleep(time.Second) // wait for some background process to cleanup
+			lenEnd := runtime.Stack(b, true)
+			// t.Log(string(b[:len]))
+			if math.Abs(float64(lenStart-lenEnd)) > 5 {
```
Comment (kevindiu) [inline policy/daemon_test.go]: how to define 5? if we add new feature the number maybe increase?

File: policy/daemon_test.go [hunk 7]
```diff
@@ -773,6 +778,142 @@ func Test_policyd_CheckPolicy(t *testing.T) {
 	}
 }
 
+func Test_policyd_CheckPolicy_goroutine(t *testing.T) {
```
Comment (kevindiu) [inline policy/daemon_test.go]: not sure if it is a good practice. as there has so many factories affecting the program runtime (GC time, memory usage, etc) making the test case meaningless... but maybe it is better to have.

File: policy/daemon_fetchPolicy_test.go (hunk 5 above)
Comment (WindzCUHK) [reply policy/daemon_fetchPolicy_test.go]: 1. test file too long, easy to crash IDE
2. too many test cases in single method and each test case is very long..., difficult to read and write

File: policy/daemon_test.go (hunk 6 above)
Comment (WindzCUHK) [reply policy/daemon_test.go]: this compare the runtime stack
sample: https://github.com/yahoojapan/athenz-authorizer/blob/bug/check-policy-multiple-role-deny-deadlock/output.txt

the memory address of the code may have 1 or 2 bytes difference,
but if there is a go routine leak, the difference will much larger

File: policy/daemon_test.go (hunk 7 above)
Comment (WindzCUHK) [reply policy/daemon_test.go]: just to test no leakage...
for TDD, add a test, fix a problem

File: policy/daemon_fetchPolicy_test.go (hunk 5 above)
Comment (kevindiu) [reply policy/daemon_fetchPolicy_test.go]: but it is harder to maintain,

File: policy/daemon_test.go (hunk 6 above)
Comment (kevindiu) [reply policy/daemon_test.go]: I think later changes maybe lead the runtime stack changes? but I think it's ok for now~

File: policy/daemon.go (hunk 3 above)
Comment (WindzCUHK) [reply policy/daemon.go]: availability V.S. correctness...

If the new policy or the cached policy is expired, it is logical to invalid all old caches...

File: policy/daemon.go (hunk 4 above)
Comment (WindzCUHK) [reply policy/daemon.go]: no comments...

File: policy/daemon.go [hunk 8]
```diff
@@ -317,11 +316,28 @@ func (p *policyd) fetchPolicy(ctx context.Context, domain string) (*SignedPolicy
...
+			glg.Errorf("Error verifying policy expiry, policy expiry: %v, valid duration: %v, valid margin: %v", sp.SignedPolicyData.Expires.Time, validDur, p.expireMargin)
+			return nil, false, errors.New("policy already expired")
+		}
+		etagValidDur = sp.SignedPolicyData.Expires.Time.Sub(fastime.Now()) - p.expireMargin
+	}
+
+	// set etag cache
+	etag := res.Header.Get("ETag")
+	if etag != "" {
+		glg.Debugf("Set domain %s with etag %v, duration: %s", domain, etag, etagValidDur)
```
Comment (kevindiu) [inline policy/daemon.go]: thinking we need to check etagValidDur <=0 before set to the cache.

File: policy/daemon.go (hunk 3 above)
Comment (kevindiu) [reply policy/daemon.go]: but this logic duplicates the logic with the etag cache?
https://github.com/yahoojapan/athenz-authorizer/pull/33/files#diff-05178a6859fde9a6cf69636df95e0aa5R333

File: policy/daemon.go (hunk 4 above)
Comment (kevindiu) [reply policy/daemon.go]: I think for safety reason it is better to return error directly...

File: policy/daemon.go (hunk 2 above)
Comment (WindzCUHK) [reply policy/daemon.go]: It depends on the user's config. If the refreshDuration > expireMargin, there will be an empty policy time gap.

Either keep this, or add the `refreshDuration > expireMargin` checking in the constructor to warn the user?

File: policy/daemon.go (hunk 3 above)
Comment (WindzCUHK) [reply policy/daemon.go]: if the return response don't have etag, the old etag cache will not be updated.

File: policy/daemon.go (hunk 2 above)
Comment (kevindiu) [reply policy/daemon.go]: no, it depends on the policy expire time, 
for example policy expire time is 1 week, refresh duration is 5 days and expire margin is 1 hour, it should work well. 
expire margin means that the margin time to force update the policy without using e-tag.

File: policy/daemon.go (hunk 3 above)
Comment (kevindiu) [reply policy/daemon.go]: For performance wise, the old etag cache shouldn't be deleted and can be used in next fetch.

File: policy/daemon.go (hunk 2 above)
Comment (kevindiu) [reply policy/daemon.go]: for your case, if the policy expire time is 1 week and user set the margin to 1 month then the error will always be returned.

Comment (WindzCUHK) [reply policy/daemon.go]: move to sp.verify()

File: policy/daemon.go (hunk 4 above)
Comment (WindzCUHK) [reply policy/daemon.go]: remove etagExpTime, return error

File: policy/daemon_fetchPolicy_test.go (hunk 5 above)
Comment (WindzCUHK) [reply policy/daemon_fetchPolicy_test.go]: combine later

File: policy/daemon.go (hunk 8 above)
Comment (WindzCUHK) [reply policy/daemon.go]: will fix

**Comment by WindzCUHK** [comment]
this PR will closes https://github.com/yahoojapan/athenz-authorizer/issues/36

File: policy/daemon.go (hunk 3 above)
Comment (WindzCUHK) [reply policy/daemon.go]: block removed

**Comment by WindzCUHK** [comment]
pending items:
1. WithPolicyExpiredDuration
	1. missing option in proxy
	1. inconsistent name
1. combine test file
1. "Update policy success with multiple athenz domains"
	- `go test ./...` => `daemon_test.go:542: policy.Update() error = role policies length is not correct, got: 0, want: 1000`

**Comment by WindzCUHK** [comment]
@kevindiu 
rename

`WithEtagFlushDuration` => `WithEtagExpiryInterval`
`WithPolicyExpiredDuration` => `WithPolicyExpiryInterval`

**Comment by kevindiu** [comment]
> @kevindiu
> rename?
> 
> `WithEtagFlushDuration` => `WithEtagExpiryPeriod`
> `WithPolicyExpiredDuration` => `WithPolicyExpiryPeriod`

Rename will affect existing users... and the new name is not that much easy to understand compared to old name.

**Comment by WindzCUHK** [comment]
@kevindiu 
`WithPolicyExpiredDuration` is not in the config.
We need  to update the config anyway.

**Comment by kevindiu** [comment]
@WindzCUHK I think follow the old naming convention `WithPolicyFlushDur` like `p.etagFlushDur` is good?

**Comment by WindzCUHK** [comment]
FlushDuration

> @WindzCUHK I think follow the old naming convention `WithPolicyFlushDur` like `p.etagFlushDur` is good?

just search on google `FlushDuration` is like the time to clean up everything.

some software use `ExpiryInterval`
https://www.ibm.com/support/knowledgecenter/SSFKSJ_8.0.0/com.ibm.mq.ref.dev.doc/q102350_.htm
but I think `Period` is the standard word for frequency...
https://en.wikipedia.org/wiki/Frequency

**Comment by kevindiu** [comment]
In golang we use duration... even I prefer interval rather than it...
https://golang.org/pkg/time/#Duration

As we use flushDur before and I think it also make sense, can we just use flushDur?

**Comment by WindzCUHK** [comment]
> In golang we use duration... even I prefer interval rather than it...
> https://golang.org/pkg/time/#Duration
> 
> As we use flushDur before and I think it also make sense, can we just use flushDur?

both works, but we better make the following consistent...
`PolicyRefreshDuration` `PubKeyErrRetryInterval`

**Comment by WindzCUHK** [comment]
maybe fix them all in other PR...?

**Comment by kevindiu** [comment]
If we want to make them all consistence then it will cause backward compatibility problem.. I think this time we can name it as policyFlushDur and change it later if we want to.

**Comment by WindzCUHK** [comment]
> If we want to make them all consistence then it will cause backward compatibility problem.. I think this time we can name it as policyFlushDur and change it later if we want to.

better to change the interface in 1 PR?
release and doc update are annoying🙈

**Comment by kevindiu** [comment]
No, if we change the name then we need to announce to existing user and lost backward compatibility. prefer keeping consistency. As flushDur also make sense.

**Comment by WindzCUHK** [comment]
fix `WithPolicyExpiredDuration` on https://github.com/yahoojapan/athenz-authorizer/issues/37

**Comment by WindzCUHK** [comment]
@kevindiu 
plx help to merge if there is no problem

File: policy/daemon.go [hunk 9]
```diff
@@ -26,6 +27,7 @@ import (
 	"sync"
 	"time"
 
+	"github.com/kpango/fastime"
```
Comment (kevindiu) [inline policy/daemon.go]: 👍🏿I forgot to use fastime

File: policy/daemon.go [hunk 10]
```diff
@@ -317,11 +316,16 @@ func (p *policyd) fetchPolicy(ctx context.Context, domain string) (*SignedPolicy
...
-	// set eTag cache
-	eTag := res.Header.Get("ETag")
-	if eTag != "" {
-		glg.Debugf("Setting ETag %v for domain %s", eTag, domain)
-		p.etagCache.SetWithExpire(domain, &etagCache{eTag, sp}, p.etagExpTime)
+	// set etag cache
+	etag := res.Header.Get("ETag")
+	if etag != "" {
+		etagValidDur := sp.SignedPolicyData.Expires.Time.Sub(fastime.Now()) - p.expireMargin
+		glg.Debugf("Set domain %s with etag %v, duration: %s", domain, etag, etagValidDur)
```
Comment (kevindiu) [inline policy/daemon.go]:
```suggestion
		glg.Debugf("set domain %s with etag %v, duration: %s", domain, etag, etagValidDur)
```

File: policy/daemon.go [hunk 11]
```diff
@@ -317,11 +316,16 @@ func (p *policyd) fetchPolicy(ctx context.Context, domain string) (*SignedPolicy
...
+		if etagValidDur > 0 {
+			p.etagCache.SetWithExpire(domain, &etagCache{etag, sp}, etagValidDur)
+		} else {
```
Comment (kevindiu) [inline policy/daemon.go]: I think it is better to have simple comment about why the cache is delete in this case

**Comment by kevindiu** [comment]
Planning to test with 
https://github.com/yahoojapan/authorization-proxy/pull/25
before I merge this 2 PR.

File: policy/signed_policy.go [hunk 12]
```diff
@@ -31,6 +34,23 @@ type SignedPolicy struct {
...
+	if s.SignedPolicyData == nil {
+		return errors.New("without policy data")
+	}
+
+	// verify expires
+	// if not set, s.SignedPolicyData.Expires is nil
+	if s.SignedPolicyData.Expires == nil {
+		return errors.New("policy without expiry")
+	}
+	// if set, but invalid, s.SignedPolicyData.Expires is Time{}
```
[the remark this answers is not in the document]
Comment (WindzCUHK) [reply policy/signed_policy.go]: The JSON decode behavior is not normal, so I changed the comment.

File: policy/daemon.go (hunk 10 above)
Comment (WindzCUHK) [reply policy/daemon.go]: some log messages start with 'A', some with 'a'.
I cannot find the pattern...

**Pull request merged**